src/mlmc/sample_storage.py

```diff
@@ -0,0 +1,134 @@
+import numpy as np
+from abc import ABCMeta
+from abc import abstractmethod
+from typing import List
+from new_simulation import QuantitySpec
+
+
+class SampleStorage(metaclass=ABCMeta):
+
+    @abstractmethod
+    def save_samples(self, correct_samples, failed_samples):
+        """
+        Write results to storag
+        """
+
+    @abstractmethod
+    def save_result_format(self, res_spec: List[QuantitySpec]):
+        """
+        Save result format
+        """
+
+    @abstractmethod
+    def load_result_format(self) -> List[QuantitySpec]:
+        """
+        Load result format
+        """
+
+    @abstractmethod
+    def sample_pairs(self):
+        """
+        Get results from storage
+        :return:
+        """
+
+
+class Memory(SampleStorage):
+
+    def __init__(self):
+        self._failed = {}
+        self._results = {}
+        self._successful_sample_ids = {}
+        self._scheduled = {}
+        self._result_specification = []
+
+    def save_samples(self, successful_samples, failed_samples):
+        """
+        Save successful samples - store result pairs
+             failed samples - store sample ids and corresponding error messages
+        :return:
+        """
+        self._save_successful(successful_samples)
+        self._save_failed(failed_samples)
+
+    def _save_successful(self, samples):
+        """
+        Save successful samples
+        :param samples: List[Tuple[sample_id: str, Tuple[ndarray, ndarray]]]
+        :return: None
+        """
+        for level_id, res in samples.items():
+            res = np.array(res)
+            fine_coarse_res = res[:, 1]
+
+            result_type = np.dtype((np.float, np.array(fine_coarse_res[0]).shape))
+            results = np.empty(shape=(len(res),), dtype=result_type)
+            results[:] = [val for val in fine_coarse_res]
+
+            # Save sample ids
+            self._successful_sample_ids.setdefault(level_id, []).extend(res[:, 0])
+
+            if level_id not in self._results:
+                self._results[level_id] = results
+            else:
+                self._results[level_id] = np.concatenate((self._results[level_id], results), axis=0)
+
+    def _save_failed(self, samples):
+        """
+        Save failed ids and error messages
+        :param samples: List[Tuple[sample_id: str, error_message: str]]
+        :return: None
+        """
+        for level_id, res in samples.items():
+            self._failed.setdefault(level_id, []).extend(res)
+
+    def save_result_format(self, res_spec: List[QuantitySpec]):
+        """
+        Save sample result format
+        :param res_spec: List[QuantitySpec]
+        :return: None
+        """
+        self._result_specification = res_spec
+
+    def n_finished(self):
```

Comment:
Possibly rename to n_successfull or introduce separate methods. Use it consistently in Sampler.